# Taxonomy

## Dataset preparation

There are three subsets of data. They correspond to filter by title, body, and tags. 
1. posts_by_title: Posts containing "Github Actions" or its variants in the question title (github actions, github-action, Github actions, Github-actions, Github Actions, Github-Actions).
2. posts_by_body: Posts containing "Github Actions" or its variants in the question body.
3. posts_by_tags: Posts tagged with any of of github actions tag list ('github-actions', 'building-github-actions', 'github-actions-self-hosted-runners', 'github-actions-runners', 'github-actions-services', 'github-actions-artifacts', 'github-actions-reusable-workflows', 'github-actions-workflows', 'github-actions-marketplace')

In [1]:
import pandas as pd
import numpy as np

In [2]:
posts_by_title = pd.read_json('data/posts_by_title.json')
posts_by_body = pd.read_json('data/posts_by_body.json')
posts_by_tags = pd.read_json('data/posts_by_tags.json')
posts_by_title.rename(columns={"CONCAT('https://stackoverflow.com/q/', p.Id)": 'link'}, inplace=True)
posts_by_body.rename(columns={"CONCAT('https://stackoverflow.com/q/', p.Id)": 'link'}, inplace=True)
posts_by_tags.rename(columns={"CONCAT('https://stackoverflow.com/q/', p.Id)": 'link'}, inplace=True)

In [3]:
posts_by_body = posts_by_body[posts_by_body['post_type_id']==1] # Keeping only post bodies that correspond to question bodies
# posts_by_tags.drop(columns=['DeletionDate', 'OwnerDisplayName', 'LastEditorDisplayName', 'FavoriteCount', 'ClosedDate', 'CommunityOwnedDate'], inplace=True);

In [4]:
# posts_by_tags.rename(columns={'Id':'id', 'PostTypeId':'post_type_id', 'AcceptedAnswerId': 'accepted_answer_id', 'ParentId': 'parent_id', 'CreationDate':'creation_date',
#                                 'Score':'score', 'ViewCount':'view_count', 'Body':'post_body', 'OwnerUserId':'owner_user_id', 'LastEditorUserId':'last_editor_user_id',
#                                 'LastEditDate':'last_edit_date', 'LastActivityDate':'last_activity_date', 'Title':'post_title', 'Tags':'tags', 'AnswerCount':'answer_count',
#                                 'CommentCount':'comment_count', 'ContentLicense':'content_license', 'Unnamed: 23':'link'}, inplace=True)


In [5]:
posts_by_tags.describe()
#posts_by_title.describe()
# posts_by_body.describe()

id  post_type_id  accepted_answer_id  parent_id        score  \
count  9.873000e+03        9873.0        3.670000e+03        0.0  9873.000000   
mean   7.126080e+07           1.0        7.022708e+07        NaN     2.747392   
std    5.184317e+06           0.0        5.424942e+06        NaN    11.666862   
min    5.417676e+07           1.0        5.417763e+07        NaN    -6.000000   
25%    6.796411e+07           1.0        6.622763e+07        NaN     0.000000   
50%    7.263306e+07           1.0        7.129186e+07        NaN     1.000000   
75%    7.562319e+07           1.0        7.493441e+07        NaN     2.000000   
max    7.759369e+07           1.0        7.758567e+07        NaN   348.000000   

          view_count  owner_user_id  last_editor_user_id  answer_count  \
count    9873.000000   9.801000e+03         4.271000e+03   9873.000000   
mean     2217.544718   8.707827e+06         6.605726e+06      0.981363   
std      8988.151347   6.554179e+06         6.039375e+06      1.119418   
min         5.000000   9.100000e+01        -1.000000e+00      0.000000   
25%       136.000000   2.628868e+06         1.623876e+06      0.000000   
50%       459.000000   7.764329e+06         4.290962e+06      1.000000   
75%      1385.000000   1.383621e+07         1.023495e+07      1.000000   
max    321529.000000   2.302257e+07         2.296569e+07     35.000000   

       comment_count  
count    9873.000000  
mean        1.657753  
std         2.394376  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         2.000000  
max        26.000000

In [6]:
df_union = pd.concat([posts_by_tags, posts_by_title, posts_by_body]).drop_duplicates().reset_index(drop=True)
df_union.describe()

id  post_type_id  accepted_answer_id  parent_id  \
count  1.132300e+04       11323.0        4.140000e+03        0.0   
mean   7.133022e+07           1.0        7.028556e+07        NaN   
std    5.145350e+06           0.0        5.374074e+06        NaN   
min    5.417676e+07           1.0        5.417763e+07        NaN   
25%    6.805283e+07           1.0        6.637378e+07        NaN   
50%    7.267653e+07           1.0        7.129904e+07        NaN   
75%    7.565271e+07           1.0        7.495300e+07        NaN   
max    7.759369e+07           1.0        7.758567e+07        NaN   

              score     view_count  owner_user_id  last_editor_user_id  \
count  11323.000000   11323.000000   1.124500e+04         4.915000e+03   
mean       2.563543    2062.098296   8.737486e+06         6.706649e+06   
std       11.054158    8502.726944   6.573781e+06         6.117352e+06   
min       -6.000000       5.000000   9.100000e+01        -1.000000e+00   
25%        0.000000     132.000000   2.628868e+06         1.611459e+06   
50%        1.000000     438.000000   7.793017e+06         4.420967e+06   
75%        2.000000    1299.000000   1.391782e+07         1.060695e+07   
max      348.000000  321529.000000   2.302257e+07         2.296569e+07   

       answer_count  comment_count  
count  11323.000000   11323.000000  
mean       0.963967       1.648238  
std        1.085275       2.396634  
min        0.000000       0.000000  
25%        0.000000       0.000000  
50%        1.000000       1.000000  
75%        1.000000       2.000000  
max       35.000000      26.000000

In [7]:
df_merged_1 = pd.merge(posts_by_tags, posts_by_title, how='inner')
df_intersection = pd.merge(df_merged_1, posts_by_body, how='inner')
df_intersection.describe()

id  post_type_id  accepted_answer_id  parent_id        score  \
count  2.903000e+03        2903.0        1.140000e+03        0.0  2903.000000   
mean   7.050473e+07           1.0        6.938052e+07        NaN     3.831898   
std    5.605103e+06           0.0        5.787363e+06        NaN    16.139532   
min    5.417676e+07           1.0        5.417763e+07        NaN    -3.000000   
25%    6.637150e+07           1.0        6.472536e+07        NaN     0.000000   
50%    7.186596e+07           1.0        7.041612e+07        NaN     1.000000   
75%    7.542059e+07           1.0        7.450144e+07        NaN     3.000000   
max    7.759154e+07           1.0        7.758567e+07        NaN   348.000000   

          view_count  owner_user_id  last_editor_user_id  answer_count  \
count    2903.000000   2.882000e+03         1.272000e+03   2903.000000   
mean     2881.829142   8.590357e+06         6.978242e+06      1.053049   
std     11830.760186   6.426077e+06         6.081297e+06      1.377158   
min         6.000000   9.100000e+01        -1.000000e+00      0.000000   
25%       183.000000   2.628868e+06         2.087967e+06      0.000000   
50%       570.000000   7.648607e+06         4.900238e+06      1.000000   
75%      1650.000000   1.352424e+07         1.106587e+07      1.000000   
max    321529.000000   2.302257e+07         2.261377e+07     35.000000   

       comment_count  
count    2903.000000  
mean        1.714433  
std         2.484013  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         3.000000  
max        26.000000

### Number of questions filtered:
- posts_by_title: 4538
- posts_by_body: 6297
- posts_by_tags: 9873
- df_intersection: 2903
- df_union: 11323

In [8]:
df_intersection['tags'].str.contains('<github-actions>').value_counts()

tags
True     2894
False       9
Name: count, dtype: int64

- We have decided to use df_intersection because we consider that we are ensuring that the topic is for sure related to GitHub Actions because it is tagged and mentioned in the title and the body question. 

- We decided to select a random sample.

- 340 posts or more are needed to have a confidence level of 95% that the real value is within ±5%.

In [9]:
# Creating the random sample. We are accepting the first 340 posts that we consider are related to the topc.
sample_size = 400
seed = 0
df_sample = df_intersection.sample(n=sample_size, random_state=seed)
print("Sample length: ", df_sample.shape[0])
df_sample.head()

Sample length:  400


id  post_type_id  accepted_answer_id  parent_id  \
582   65011370             1                 NaN        NaN   
1811  73822327             1          73822678.0        NaN   
2249  75607496             1                 NaN        NaN   
1652  73018888             1                 NaN        NaN   
667   65855054             1                 NaN        NaN   

            creation_date  score  view_count  \
582   2020-11-25 19:17:14      0         637   
1811  2022-09-23 02:09:56      1         132   
2249  2023-03-01 18:12:01      1         107   
1652  2022-07-18 07:48:42      2        1437   
667   2021-01-23 02:10:44      3        5595   

                                              post_body  owner_user_id  \
582   <p>I have this go.yml for github actions</p>&#...     12215821.0   
1811  <p>I can locally create and use a docker conta...      7628816.0   
2249  <p>We have a legacy Maven application which is...      3065868.0   
1652  <p>GitHub Actions concurrency broke my process...       119790.0   
667   <p>So I have a repo with multiple directories ...      3681199.0   

      last_editor_user_id       last_edit_date   last_activity_date  \
582                   NaN                 None  2020-11-26 08:31:42   
1811            7628816.0  2022-09-23 03:25:16  2022-09-23 03:25:37   
2249            3065868.0  2023-03-03 08:57:38  2023-03-03 08:57:38   
1652                  NaN                 None  2022-10-03 23:31:28   
667                   NaN                 None  2021-01-23 08:11:48   

                                             post_title  \
582         Github actions, problem with dep installing   
1811       ory/dockertest not working on GitHub Actions   
2249  GitHub Actions: Generate Deployment sources fo...   
1652          How to disable GitHub Actions Concurrency   
667   Can you have multiple working directories with...   

                                                   tags  answer_count  \
582                           <go><github-actions><dep>             1   
1811          <docker><github-actions><clickhouse><ory>             1   
2249  <maven><github-actions><websphere><websphere-8...             0   
1652                           <github><github-actions>             1   
667                                <go><github-actions>             1   

      comment_count content_license                                  link  
582               0    CC BY-SA 4.0  https://stackoverflow.com/q/65011370  
1811              0    CC BY-SA 4.0  https://stackoverflow.com/q/73822327  
2249              9    CC BY-SA 4.0  https://stackoverflow.com/q/75607496  
1652              2    CC BY-SA 4.0  https://stackoverflow.com/q/73018888  
667               2    CC BY-SA 4.0  https://stackoverflow.com/q/65855054

In [10]:
df_sample[['id', 'link']].assign(evaluator1=np.nan).to_excel('data/sample_eval1.xlsx')
df_sample[['id', 'link']].assign(evaluator2=np.nan).to_excel('data/sample_eval2.xlsx')

In [11]:
sample_reviewed=pd.read_excel('data/sample_eval1_reviewed.xlsx')
sample_reviewed['evaluator2']=pd.read_excel('data/sample_eval2_reviewed.xlsx')['evaluator2']

In [12]:
sample_accepted=sample_reviewed[(sample_reviewed['evaluator1']==1)&(sample_reviewed['evaluator2']==1)].head(340)
sample_accepted = pd.merge(df_intersection, sample_accepted[['id']], on='id', how='inner')
sample_accepted.to_excel('data/sample_accepted.xlsx')

In [13]:
sample_accepted

id  post_type_id  accepted_answer_id  parent_id  \
0    57503578             1                 NaN        NaN   
1    57509118             1          57521953.0        NaN   
2    57639507             1          57639714.0        NaN   
3    57806624             1          57806894.0        NaN   
4    57808152             1          59022667.0        NaN   
..        ...           ...                 ...        ...   
335  77395114             1          77395417.0        NaN   
336  77431693             1                 NaN        NaN   
337  77446605             1          77446960.0        NaN   
338  77469265             1                 NaN        NaN   
339  77519735             1                 NaN        NaN   

           creation_date  score  view_count  \
0    2019-08-15 00:30:14     13        7788   
1    2019-08-15 11:42:14      4        2302   
2    2019-08-24 16:02:24      4         819   
3    2019-09-05 13:29:41     80       35589   
4    2019-09-05 14:56:14     11        4658   
..                   ...    ...         ...   
335  2023-10-31 10:30:28      0          62   
336  2023-11-06 14:04:11      0          32   
337  2023-11-08 14:49:09      1          95   
338  2023-11-12 14:11:32      0          28   
339  2023-11-21 00:52:54      1         304   

                                             post_body  owner_user_id  \
0    <p>I have a file in a GitHub repository that n...        54929.0   
1    <p>I am working with new GitHub actions, idea ...       911930.0   
2    <p>I'm trying to set up a CI/CD pipeline in Gi...      1341734.0   
3    <p>I'm using GitHub Actions to build my projec...      3231778.0   
4    <p>I'm trying out GitHub Actions to build my F...      3231778.0   
..                                                 ...            ...   
335  <p>I wanted to experiment with GitHub Actions:...     11186407.0   
336  <p>Basically as it says in the title.</p>&#xA;...     22436104.0   
337  <p>I have my unittests in a top level <code>te...       471136.0   
338  <p>I'm facing an issue with accessing metadata...     22902806.0   
339  <p>I'm uploading the changes in NestJS via Git...     20892191.0   

     last_editor_user_id       last_edit_date   last_activity_date  \
0                    NaN                 None  2023-04-05 01:42:41   
1                    NaN                 None  2019-08-16 09:15:48   
2                    NaN                 None  2022-04-06 11:40:39   
3                    NaN                 None  2022-03-13 19:24:54   
4              3231778.0  2019-09-05 15:48:37  2022-09-14 10:01:40   
..                   ...                  ...                  ...   
335                  NaN                 None  2023-10-31 13:15:41   
336            9938317.0  2023-11-06 15:35:14  2023-11-06 15:35:14   
337                  NaN                 None  2023-11-08 15:41:57   
338                  NaN                 None  2023-11-12 14:11:32   
339            3001761.0  2023-11-21 08:01:48  2023-11-21 08:01:48   

                                            post_title  \
0    Making pull requests to a GitHub repository au...   
1              New GitHub actions run in empty folders   
2     How to access a service in Github Actions CI/CD?   
3    GitHub Actions - How to build project in sub-d...   
4         How to build Flutter in GitHub Actions CI/CD   
..                                                 ...   
335  GitHub Actions generated docker image results ...   
336  GitHub Actions VS code extension gives access ...   
337  Running Python Poetry unit test in Github actions   
338  How to Retrieve Metadata of AWS-Stored Audio F...   
339    Dependency installation error in GitHub Actions   

                                                  tags  answer_count  \
0                             <github><github-actions>             1   
1                  <github><github-actions><github-ci>             1   
2                     <github><elixir><github-actions>             1  

In [14]:
from bs4 import BeautifulSoup
import re

In [21]:
def html_to_sentences_df(html_content):
    """
    Parses the provided HTML content, replaces specific tags with placeholders,
    adjusts paragraph endings, and splits the content into sentences.
    Returns a pandas DataFrame with each sentence in a separate row.
    
    Parameters:
    - html_content: String containing HTML content.
    
    Returns:
    - DataFrame with each sentence as a separate row.
    """
    
    # Parse the HTML
    soup = BeautifulSoup(html_content, 'html.parser')

    # Replace code blocks, blockquotes, and links with placeholders
    for code in soup.find_all('code'):
        code.replace_with("-CODE-")
    for blockquote in soup.find_all('blockquote'):
        blockquote.replace_with("-BLOCK-")
    for a in soup.find_all('a'):
        a.replace_with("-LINK-")

    # Extract text and replace newline entities
    text = soup.get_text()
    text = text.replace('&#xA;', '\n').strip()

    # Pre-process text to handle ':\\n-CODE-' pattern
    text = re.sub(r':\s*\n-CODE-', ': -CODE-', text)
    text = re.sub(r':\s*\n-BLOCK-', ': -BLOCK-', text)
    text = re.sub(r':\s*\n-LINK-', ': -LINK-', text)
    
    # Adjust paragraph endings where necessary
    pattern = r'(?<![\.\!\?\s])\s*\n'
    text = re.sub(pattern, '.\n', text)

    # Replace ':.' with ':'
    text = text.replace(':.', ':')

    # Split text into sentences
    sentences = re.split(r'(?<=[.!?]) +', text.replace('\n', ' '))

    # Create DataFrame
    df = pd.DataFrame(sentences, columns=['sentence'])
    
    return df

In [23]:
def process_df(sample_df):
    # Initialize an empty DataFrame to hold all sentences
    all_sentences_df = pd.DataFrame()

    # Iterate over each row in the DataFrame
    for index, row in sample_df.iterrows():

        # Process the title and body, assuming they are HTML content
        title_sentences_df = html_to_sentences_df(row['post_title'])
        body_sentences_df = html_to_sentences_df(row['post_body'])

        # Add a column with the index/id of the post
        title_sentences_df['id'] = index
        body_sentences_df['id'] = index

        # Add a column with the index/id of the post
        title_sentences_df['post_id'] = row['id']
        body_sentences_df['post_id'] = row['id']

        # Add a column to indicate the source of the sentences
        title_sentences_df['source'] = 'title'
        body_sentences_df['source'] = 'body'
        
        # Combine title and body sentences
        combined_sentences_df = pd.concat([title_sentences_df, body_sentences_df], ignore_index=True)
        
        # Add the combined sentences to the overall DataFrame
        all_sentences_df = pd.concat([all_sentences_df, combined_sentences_df], ignore_index=True)

        all_sentences_df = all_sentences_df[['id', 'post_id', 'source', 'sentence']]
    
    return all_sentences_df

all_sentences_df = process_df(sample_accepted)

C:\Users\shern\AppData\Local\Temp\ipykernel_44400\519759939.py:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  soup = BeautifulSoup(html_content, 'html.parser')
C:\Users\shern\AppData\Local\Temp\ipykernel_44400\519759939.py:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  soup = BeautifulSoup(html_content, 'html.parser')
C:\Users\shern\AppData\Local\Temp\ipykernel_44400\519759939.py:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  soup = BeautifulSoup(html_content, 'html.parser')
C:\Users\shern\AppData\Local\Temp\ipykernel_44400\519759939.py:15: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open thi

In [17]:
categories = [1, 4, 5, 7, 2, 9, 13, 11, 14, 20, 22, 19, 23]
# Add each new column filled with zeros to the DataFrame
for column in categories:
    all_sentences_df[column] = 0

In [18]:
all_sentences_df.to_excel('data/sentences_eval1.xlsx')
all_sentences_df.to_excel('data/sentences_eval2.xlsx')

In [45]:
classified_sentences_eval1 = pd.read_excel('data/sentences_eval1_reviewed.xlsx')
classified_sentences_eval2 = pd.read_excel('data/sentences_eval2_reviewed.xlsx')

In [46]:
classified_sentences_eval1.columns.drop(['Unnamed: 0', 'id', 'post_id', 'source', 'sentence'])

Index([1, 4, 5, 7, 2, 9, 13, 11, 14, 20, 22, 19, 23], dtype='object')

In [48]:
import pandas as pd
from sklearn.metrics import cohen_kappa_score


kappa = pd.DataFrame()
for category in classified_sentences_eval1.columns.drop(['Unnamed: 0', 'id', 'post_id', 'source', 'sentence']):
    kappa[category] = [cohen_kappa_score(classified_sentences_eval1[category], classified_sentences_eval2[category])]
    
kappa

1         4         5         7    2         9    13       11   14  \
0  0.439652  0.238585  0.240824  0.219126  0.0  0.405172  0.0  0.05899  0.0   

         20        22   19   23  
0  0.498816  0.362615  0.0  0.0